File: ai-app-week3/frontend/src/api/embeddings.ts
const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || "http://127.0.0.1:8010";

export interface ChunkEmbeddingResponse {
  id: number;
  chunk_id: number;
  provider: string;
  model: string;
  dimension: number;
  embedding: number[];
  status: string;
  error_message?: string | null;
  created_at: string;
  updated_at: string;
}

async function parseError(response: Response, fallbackMessage: string) {
  try {
    const data = await response.json();
    return data?.detail || fallbackMessage;
  } catch {
    return fallbackMessage;
  }
}

export async function createChunkEmbedding(
  chunkId: number
): Promise<ChunkEmbeddingResponse> {
  const response = await fetch(
    `${API_BASE_URL}/ai/embeddings/chunks/${chunkId}`,
    {
      method: "POST",
    }
  );

  if (!response.ok) {
    const message = await parseError(
      response,
      `创建 chunk embedding 失败：${response.status}`
    );
    throw new Error(message);
  }

  return response.json();
}

export async function getChunkEmbedding(
  chunkId: number
): Promise<ChunkEmbeddingResponse> {
  const response = await fetch(
    `${API_BASE_URL}/ai/embeddings/chunks/${chunkId}`,
    {
      method: "GET",
    }
  );

  if (!response.ok) {
    const message = await parseError(
      response,
      `获取 chunk embedding 失败：${response.status}`
    );
    throw new Error(message);
  }

  return response.json();
}

export interface BatchEmbeddingItem {
  chunk_id: number;
  status: "success" | "skipped" | "failed";
  message: string;
  embedding_id?: number | null;
  dimension?: number;
}

export interface BatchDocumentEmbeddingResponse {
  document_id: number;
  total_active_chunks: number;
  success_count: number;
  skipped_count: number;
  failed_count: number;
  items: BatchEmbeddingItem[];
}

export async function createDocumentChunkEmbeddings(
  documentId: number,
  skipExisting = true
): Promise<BatchDocumentEmbeddingResponse> {
  const response = await fetch(
    `${API_BASE_URL}/ai/embeddings/documents/${documentId}/chunks?skip_existing=${skipExisting}`,
    {
      method: "POST",
    }
  );

  if (!response.ok) {
    const message = await parseError(
      response,
      `批量创建文档 embeddings 失败：${response.status}`
    );
    throw new Error(message);
  }

  return response.json();
}

export interface EmbeddingSearchRequest {
  query: string;
  top_k: number;
  document_id?: number | null;
  only_active: boolean;
  quality_status?: string | null;
  score_threshold?: number | null;
}

export interface EmbeddingSearchResultItem {
  chunk_id: number;
  document_id?: number | null;
  embedding_id: number;
  score: number;
  content: string;
  content_preview: string;
  provider: string;
  model: string;
  dimension: number;
  is_active?: boolean | null;
  status?: string | null;
  quality_status?: string | null;
  quality_note?: string | null;
  error_message?: string;
}

export interface EmbeddingSearchResponse {
  query: string;
  top_k: number;
  document_id?: number | null;
  only_active: boolean;
  quality_status?: string | null;
  score_threshold?: number | null;
  applied_filters: {
    document_id?: number | null;
    only_active: boolean;
    quality_status?: string | null;
    score_threshold?: number | null;
  };
  total_candidates: number;
  returned_count: number;
  matched_after_score_filter: number;
  max_score?: number | null;
  min_score?: number | null;
  results: EmbeddingSearchResultItem[];
}

export async function searchChunkEmbeddings(
  payload: EmbeddingSearchRequest
): Promise<EmbeddingSearchResponse> {
  const response = await fetch(`${API_BASE_URL}/ai/embeddings/search`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(payload),
  });

  if (!response.ok) {
    const message = await parseError(
      response,
      `向量检索失败：${response.status}`
    );
    throw new Error(message);
  }

  return response.json();
}

File: ai-app-week3/backend/app/routers/embeddings.py
import json
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from app.database import get_db
from app.services.chunk_embedding_service import ChunkEmbeddingService

from pydantic import BaseModel, Field

from app.services.embedding_search_service import EmbeddingSearchService

from typing import Optional

router = APIRouter(prefix="/ai/embeddings", tags=["embeddings"])


@router.post("/chunks/{chunk_id}")
def embed_chunk(chunk_id: int, db: Session = Depends(get_db)):
    service = ChunkEmbeddingService(db)

    try:
        record = service.embed_chunk(chunk_id)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))

    return {
        "id": record.id,
        "chunk_id": record.chunk_id,
        "provider": record.provider,
        "model": record.model,
        "dimension": record.dimension,
        "embedding": json.loads(record.embedding_json),
        "status": record.status,
        "error_message": record.error_message,
        "created_at": record.created_at,
        "updated_at": record.updated_at,
    }


@router.get("/chunks/{chunk_id}")
def get_chunk_embedding(chunk_id: int, db: Session = Depends(get_db)):
    service = ChunkEmbeddingService(db)

    record = service.get_chunk_embedding(chunk_id)

    if not record:
        raise HTTPException(status_code=404, detail="当前 chunk 还没有 embedding")

    return {
        "id": record.id,
        "chunk_id": record.chunk_id,
        "provider": record.provider,
        "model": record.model,
        "dimension": record.dimension,
        "embedding": json.loads(record.embedding_json),
        "status": record.status,
        "error_message": record.error_message,
        "created_at": record.created_at,
        "updated_at": record.updated_at,
    }

@router.post("/documents/{document_id}/chunks")
def embed_document_chunks(
    document_id: int,
    skip_existing: bool = True,
    db: Session = Depends(get_db),
):
    service = ChunkEmbeddingService(db)

    try:
        result = service.embed_document_chunks(
            document_id=document_id,
            skip_existing=skip_existing,
        )
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))

    return result


from app.schemas import EmbeddingSearchRequest, EmbeddingSearchResponse
from app.services.embedding_search_service import search_similar_chunks

router = APIRouter(prefix="/ai/embeddings", tags=["embeddings"])


@router.post("/search", response_model=EmbeddingSearchResponse)
def search_embeddings(
    payload: EmbeddingSearchRequest,
    db: Session = Depends(get_db),
):
    return search_similar_chunks(
        db=db,
        query=payload.query,
        document_id=payload.document_id,
        top_k=payload.top_k,
    )

class EmbeddingSearchRequest(BaseModel):
    query: str = Field(..., min_length=1, description="用户检索问题")
    top_k: int = Field(default=5, ge=1, le=20)
    document_id: Optional[int] = None
    only_active: bool = True
    quality_status: Optional[str] = Field(
        default=None,
        description="可选：unknown / good / needs_review / bad",
    )
    score_threshold: Optional[float] = Field(
        default=None,
        ge=0,
        le=1,
        description="最低相似度分数过滤，例如 0.3",
    )

@router.post("/search")
def search_embeddings(
    payload: EmbeddingSearchRequest,
    db: Session = Depends(get_db),
):
    try:
        result = search_similar_chunks(
            db=db,
            query=payload.query,
            top_k=payload.top_k,
            document_id=payload.document_id,
            only_active=payload.only_active,
            quality_status=payload.quality_status,
            score_threshold=payload.score_threshold,
        )
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))

    return result